odl/ufunc_ops/ufunc_ops.py

```diff
@@ -0,0 +1,359 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Ufunc operators for ODL vectors."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+import numpy as np
+from odl.set import LinearSpace, RealNumbers, Field
+from odl.space import ProductSpace, fn
+from odl.operator import Operator, MultiplyOperator
+from odl.util.utility import is_int_dtype
+from odl.util.ufuncs import UFUNCS
+from odl.solvers import (Functional, ScalingFunctional, FunctionalQuotient,
+                         ConstantFunctional, IdentityFunctional)
+
+__all__ = ()
+
+
+def _is_integer_only_ufunc(name):
+    return 'shift' in name or 'bitwise' in name or name == 'invert'
+
+LINEAR_UFUNCS = ['negative', 'rad2deg', 'deg2rad', 'add', 'subtract']
+
+
+RAW_EXAMPLES_DOCSTRING = """
+Examples
+--------
+>>> import odl
+>>> space = odl.{space!r}
+>>> op = odl.ufunc_ops.{name}(space)
+>>> print(op({arg}))
+{result}
+"""
+
+
+def gradient_factory(name):
+    """Create gradient `Functional` for some ufuncs."""
+
+    if name == 'sin':
+        def gradient(self):
+            """Return the gradient operator."""
+            return cos(self.domain)
+    elif name == 'cos':
+        def gradient(self):
+            """Return the gradient operator."""
+            return -sin(self.domain)
+    elif name == 'tan':
+        def gradient(self):
+            """Return the gradient operator."""
+            return 1 + square(self.domain) * self
+    elif name == 'sqrt':
+        def gradient(self):
+            """Return the gradient operator."""
+            return FunctionalQuotient(ConstantFunctional(self.domain, 0.5),
+                                      self)
+    elif name == 'square':
+        def gradient(self):
+            """Return the gradient operator."""
+            return ScalingFunctional(self.domain, 2.0)
+    elif name == 'log':
+        def gradient(self):
+            """Return the gradient operator."""
+            return FunctionalQuotient(ConstantFunctional(self.domain, 1.0),
+                                      IdentityFunctional(self.domain))
+    elif name == 'exp':
+        def gradient(self):
+            """Return the gradient operator."""
+            return self
```

Comment:
Cool, I like the simple gradient formulas.
Further simple additions are `cosh` and `sinh` since they are each other's derivatives. Also `tanh` with derivative `1/tanh^2`.
I think also `reciprocal` would be very useful, you could use it for the `tanh` thing.